torchx/distributed/__init__.py

```diff
@@ -0,0 +1,261 @@
+            "  1) https://pytorch.org/torchx/latest/components/distributed.html\n"
+            "  2) https://pytorch.org/docs/stable/elastic/run.html\n"
+            "=============================================================================================="
+        )
+        return 0
+    else:
+        return int(os.environ["LOCAL_RANK"])
+
+
+def local_cuda_device() -> torch.device:
+    """
+    Returns the CUDA device (as a ``torch.device``) based on the local rank.
+
+    See Also: :py:func:`get_local_rank`.
+    """
+    return torch.device(f"cuda:{local_rank()}")
+
+
+def rank() -> int:
+    """
+    A non-distributed-safe get_rank call. Unlike ``torch.distributed.get_rank()``
+    this method will not fail if being invoked from a non-distributed (e.g. process group not initialized)
+    context. Therefore, this method is safe to use in internal methods that may be used
+    in non-distributed contexts as well.
+
+    Returns:
+        If a process group has been initialized returns the value returned by ``torch.distributed.get_rank()``.
+        Otherwise, returns 0 (trivial rank)
+
+    """
+    return dist.get_rank() if dist.is_initialized() else 0
+
+
+def world_size() -> int:
+    """
+    A non-distributed-safe get_world_size call. Unlike ``torch.distributed.get_world_size()``,
+    this method will not fail if being invoked from a non-distributed (e.g. process group not initialized)
+    context. Threefore, this method is safe to use in internal mthods that may be used
+    in non-distributed contexts as well.
+
+    Returns:
+        If a process group has been initialized returns the value returns by ``torch.distributed.get_world_size()``.
+        Otherwise, returns 1 (trivial world_size)
+
+    """
+    return dist.get_world_size() if dist.is_initialized() else 1
+
+
+def is_rank0() -> bool:
+    """
+    Returns ``True`` if the caller is rank 0 (in a distributed setting).
+    If no process group has been initialized, then this method assumes
+    that the caller is a single-process (aka not-distributed) and trivially returns ``True``.
+    That is, for a non-distributed job there is only one process and hence that process
+    is trivially rank 0.
+
+    .. note::
+      To initialize the process group prefer to use :py:func:init_process_group over
+      ``torch.distributed.init_process_group()`` since the former can be called from
+      both distributed and non-distributed scripts.
+
+    """
+    return rank() == 0
+
+
+def is_local_rank0() -> bool:
+    """
+    Returns ``True`` if this process is local rank 0 and ``False`` otherwise.
+    Used to perform an action just once per node. Example
+
+    .. code-block:: python
+
+     if is_local_rank0():
+        # download a file just once per node
+        download_file("s3://...")
+
+
+    """
+    return local_rank() == 0
+
+
+Backend = Literal["nccl", "gloo", "auto"]
+
+
+def init_pg(backend: Backend = "auto", **kwargs: Any) -> torch.device:
```

Comment:
is this going away with 2.0?

Reply:
kind of, but not exactly the way its done here. In 2.0 if you create a pg without a backend, it'll run collectives on cpu or gpu dynamically based on the tensor's device. What I want here is to chose (for all tensors) a backend cpu or gpu (gloo vs nccl) upfront based on the availability of CUDA devices.